File: sdks/typescript/packages/core/src/proto/diagnostics.ts
/**
 * Protocol diagnostic types for client-side observability.
 *
 * The reader loop emits diagnostics whenever it cannot route a response
 * line: malformed JSON, missing id, unknown id, or non-routable protocol
 * errors. Diagnostics are buffered in a bounded ring (32 entries) and
 * trigger a desync error when the rate exceeds the configured threshold.
 */

export type ProtocolDiagnosticKind =
  | "malformed_json"
  | "non_object_response"
  | "invalid_jsonrpc_version"
  | "missing_id"
  | "unknown_id"
  | "non_routable_error";

export interface ProtocolDiagnostic {
  readonly kind: ProtocolDiagnosticKind;
  readonly message: string;
  readonly rawLine: string;
  readonly timestampMs: number;
}

export interface ProtocolLogger {
  warn(message: string, ...args: unknown[]): void;
  error(message: string, ...args: unknown[]): void;
}

const DEFAULT_BUFFER_SIZE = 32;
const MAX_LINE_PREVIEW = 512;

export class ProtocolDiagnosticBuffer {
  private readonly capacity: number;
  private readonly entries: ProtocolDiagnostic[] = [];

  constructor(capacity: number = DEFAULT_BUFFER_SIZE) {
    if (capacity <= 0) {
      throw new Error(
        `ProtocolDiagnosticBuffer capacity must be positive, got ${capacity}`,
      );
    }
    this.capacity = capacity;
  }

  push(diagnostic: ProtocolDiagnostic): void {
    if (this.entries.length >= this.capacity) {
      this.entries.shift();
    }
    this.entries.push(diagnostic);
  }

  snapshot(): readonly ProtocolDiagnostic[] {
    return this.entries.slice();
  }

  countWithin(windowMs: number, nowMs: number): number {
    const cutoff = nowMs - windowMs;
    let count = 0;
    for (let i = this.entries.length - 1; i >= 0; i -= 1) {
      if (this.entries[i]!.timestampMs >= cutoff) {
        count += 1;
      } else {
        break;
      }
    }
    return count;
  }

  clear(): void {
    this.entries.length = 0;
  }
}

export function previewLine(line: string): string {
  if (line.length <= MAX_LINE_PREVIEW) return line;
  return line.slice(0, MAX_LINE_PREVIEW) + "…";
}

File: sdks/python/src/attest/_proto/diagnostics.py
"""Protocol diagnostics for client-side observability.

The reader loop emits a ``ProtocolDiagnostic`` whenever it cannot route a
response line (malformed JSON, missing id, unknown id, or non-routable
protocol error). Diagnostics are buffered in a bounded ring (32 entries
by default). When more than ``desync_threshold`` diagnostics appear
within ``desync_window_seconds`` the client raises
:class:`ProtocolDesyncError` to fail every in-flight request.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Literal, Protocol

ProtocolDiagnosticKind = Literal[
    "malformed_json",
    "non_object_response",
    "invalid_jsonrpc_version",
    "missing_id",
    "unknown_id",
    "non_routable_error",
]

DEFAULT_BUFFER_SIZE: int = 32
DEFAULT_DESYNC_THRESHOLD: int = 3
DEFAULT_DESYNC_WINDOW_SECONDS: float = 1.0
MAX_LINE_PREVIEW: int = 512


@dataclass(frozen=True, slots=True)
class ProtocolDiagnostic:
    """A single observability record for a non-routable protocol line."""

    kind: ProtocolDiagnosticKind
    message: str
    raw_line: str
    timestamp: float


class ProtocolLogger(Protocol):
    """Minimal logger protocol used by AttestClient.

    Compatible with :class:`logging.Logger` and ad-hoc duck-typed objects.
    """

    def warning(self, msg: str, *args: object) -> None: ...
    def error(self, msg: str, *args: object) -> None: ...


class ProtocolDiagnosticBuffer:
    """Bounded ring buffer for protocol diagnostics."""

    def __init__(self, capacity: int = DEFAULT_BUFFER_SIZE) -> None:
        if capacity <= 0:
            raise ValueError(f"ProtocolDiagnosticBuffer capacity must be positive, got {capacity}")
        self._entries: deque[ProtocolDiagnostic] = deque(maxlen=capacity)

    def push(self, diagnostic: ProtocolDiagnostic) -> None:
        self._entries.append(diagnostic)

    def snapshot(self) -> tuple[ProtocolDiagnostic, ...]:
        return tuple(self._entries)

    def count_within(self, window_seconds: float, now: float) -> int:
        cutoff = now - window_seconds
        count = 0
        for entry in reversed(self._entries):
            if entry.timestamp >= cutoff:
                count += 1
            else:
                break
        return count

    def clear(self) -> None:
        self._entries.clear()


def preview_line(line: str) -> str:
    """Truncate long lines for safer logging and storage."""
    if len(line) <= MAX_LINE_PREVIEW:
        return line
    return line[:MAX_LINE_PREVIEW] + "…"
